src/Producer.py

```diff
@@ -0,0 +1,100 @@
+"""
+<Program Name>
+  Producer
+
+<Started>
+  July 2019
+
+<Author>
+  Junjie Ge
+
+<Purpose>
+  Instance of posix_omni_parser that poputales the syscall_object and trace in
+  trace manager
+
+"""
+
+import logging
+import sys
+
+from posix_omni_parser.parsers.StraceParser import StraceParser
+
+import consts
+
+class Producer:
+  def __init__(self, trace_file, pickle_file, tm):
+    # location of tracefile
+    self.tracefile = trace_file
+    self.trace_manager = tm
+    self.parser = StraceParser(self.tracefile, pickle_file)
+
+
+  def produce(self, thread_condition, backlog_size=None):
+    """
+    <Purpose>
+      This method parses and adds syscalls to the list of syscalls in
+      TraceManager. It works like a sliding window that constantly updates the
+      list depending on where the slowest mutator is in that list
+
+    <Returns>
+      None
+
+    """
+
+    if not backlog_size:
+      backlog_size = consts.BACKLOG_SIZE
+
+    with open(self.tracefile, 'r') as fh:
+      # Finding and ignoring everything before 'syscall_'
+      while True:
+        try:
+          trace_line = fh.next()
+        except StopIteration:
+          logging.debug("Incomplete Trace. Trace ended without 'syscall_'")
+          raise RuntimeError
+        syscall = self.parser.parse_line(trace_line)
+        if syscall and 'syscall_' in syscall.name:
+          break
+
+      # before updating and deleting the syscall_objects simultaneously, first add an
+      # intial amount of syscall_objects to the trace_manager in order for
+      # mutators to have an initial amout of backlogs
+      for i in range(backlog_size * 2):
+        with thread_condition:
+          try:
+            trace_line = fh.next()
+          except StopIteration:
+            self.trace_manager.producer_done()
```

Comment:
Based on my understanding of how the conditions work you're in an exclusive zone here right?  That is, the trace manager can't be checking its producer_running field while the producer has thread_condition?  Otherwise, this is a race condition.